### DISCUSSION TITLE
Best way to clean a provider with all dependencies (e.g. after log out)

### DISCUSSION THREAD

**Discussion by trejdych** [rrousselGit/riverpod#1436] [Q&A]
Hello 👋 

how can I dispose/clean some providers without immediately recreate? 
e.g. I want have some user specific request (personal data/comments/etc) 
``` dart
final authorizationStateProvider = StateProvider<bool>((ref) => false);

final someRequestProvider = FutureProvider<String>((ref) async {
  // do some request
  return "some result";
});

final someOtherRequestProvider = FutureProvider<String>((ref) async {
  final someRequest = await ref.watch(someRequestProvider.future);
  // do some other request
  return "$someRequest other result";
});
```

I can do something like that:

``` dart
final someRequestProvider = FutureProvider<String>((ref) async {
  if (!ref.watch(authorizeStateProvider)) {
      throw const ApiException.unauthorizedException();
    }
  // do some request
  return "some result";
});
```
and it is ok for some cases but if I need someRequestProvider for non logged user it can't work 🙂  on other hand I don't want to do sth like that
``` dart
final someRequestProvider = FutureProvider<String>((ref) async {
  ref.watch(authorizeStateProvider));
  // do some request
  return "some result";
}); 
// or
final someRequestProvider = FutureProvider<String>((ref) async {
  ref.listen<bool>(
    authorizationStateProvider,
    (previous, next) {
      if (previous != null && previous != next) {
        ref.invalidateSelf();
      }
    },
  );
  // do some request
  return "some result";
});
```
because I don't need  that request immediately after logout.

Could you help me? 🙏

**Comment by rrousselGit** [comment]
Could you ewplain more about what you're trying to do ewactly? The true problem you're facing is unclear to me. I don't understand how the code you gave relates to the title of this discussion

**Reply by trejdych** [reply]
I'll try 🙂 Relation between the code and the title is that I want to dispose (without re-creating) `someRequestProvider` and `someOtherRequestProvider` after `authorizationStateProvider change

so I want to get sth like that:
1. user log in (authorizationState == true) and goes to the screen with `someRequestProvider` or/and `someOtherRequestProvider`
2. the user got a result, and then the user log out (authorizationState == false) and goes to the login screen
3. I want both my FutureProviders to be disposed

**Reply by trejdych** [reply]
After changing to autoDispose and using keepAliveLink I got what I wanted but I'm not convinced of that solution. 

``` dart
final someRequestProvider = FutureProvider.autoDispose<String>((ref) async {
  // do some request
  final keepAliveLink = ref.keepAlive();
  ref.listen<bool>(
    authorizationStateProvider,
    (previous, next) {
      if (previous != null && previous != next) {
        keepAliveLink.close();
      }
    },
  );
  return "some result";
});

final someOtherRequestProvider =
    FutureProvider.autoDispose<String>((ref) async {
  final someRequest = await ref.watch(someRequestProvider.future);
  return "$someRequest other result";
});

```

btw. Grammarly says that my text may sound like dart 🎯 💪 
<img width="210" alt="CleanShot 2022-06-02 at 09 44 25@2x" src="https://user-images.githubusercontent.com/4202808/171580308-50976f69-df39-4bc9-8eae-6f61e640d18f.png">
